Source/NSubstitute/Core/CallSpecificationFactory.cs

```diff
@@ -15,7 +15,7 @@ public ICallSpecification CreateFrom(ICall call, MatchArgs matchArgs)
         {
             var methodInfo = call.GetMethodInfo();
             var argumentSpecs = call.GetArgumentSpecifications();
-            var arguments = call.GetArguments();
+            var arguments = call.GetOriginalArguments();
```

Comment:
@dtchepak While Alexandr is checking that, do you see from your side any problems that could arise because of this change? The idea is that we always use original call arguments for specifications, rather than potentially modified args. So if user changes any argument value (e.g. in `When/Do` handler), that would not affect call specification.